dynamic_programming/abbreviation.cpp

```diff
@@ -0,0 +1,164 @@
+ */
+
+#include <cassert>
+#include <iostream>
+#include <string>
+#include <vector>
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming Algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @namespace Knapsack
+ * @brief Implementation of Abbreivation problem
+ */
+namespace abbreviation {
+/**
+ *  (recursive dp function) https://www.hackerrank.com/challenges/abbr/problem
+ *  Returns whether s can be converted to t with following rules:
+ *  a. Capitalize zero or more of a's lowercase letters from string s
+ *  b. remove all other lowercase letters from string s
+ * @param dp: memo as parameter to store the result
+ * @param v: visited boolean to check if the result is already computed
+ * @param s: given string
+ * @param t: resultant abbreivated string
+ * @param i: start of string s
+ * @param j: start of string j
+ * @returns bool whether s can be converted to t
+ */
+bool abbreviation_recursion(std::vector<std::vector<bool>> *memo,
+                            std::vector<std::vector<bool>> *visited,
+                            const std::string &s, const std::string &t,
+                            int i = 0, int j = 0) {
+    bool ans = memo->at(i).at(j);
+    if (i == s.size() && j == t.size()) {
+        return true;
+    } else if (i == s.size() && j != t.size()) {
+        // result t is not converted, return false
+        return false;
+    } else if (!visited->at(i).at(j)) {
+        /**
+         * (s[i] == t[j]): if s char at position i is equal to t char at
+         * position j, then s character is a capitalized one, move on to next
+         * character
+         * s[i] - 32 == t[j]: if s[i] character is lowercase of t[j] then
+         * explore two possibilites:
+         * 1. convert it to capitalized and move both to next pointer (i + 1, j
+         * + 1)
+         * 2. Discard the character (s[i]) and move to next char (i + 1, j)
+         */
+        if (s[i] == t[j]) {
+            ans = abbreviation_recursion(memo, visited, s, t, i + 1, j + 1);
+        } else if (s[i] - 32 == t[j]) {
+            ans = abbreviation_recursion(memo, visited, s, t, i + 1, j + 1) ||
+                  abbreviation_recursion(memo, visited, s, t, i + 1, j);
+        } else {
+            // if s[i] is uppercase, then cannot be converted, return false
+            // else s[i] is lowercase, only option is to discard this character
+            if (s[i] >= 'A' && s[i] <= 'Z') {
+                ans = false;
+            } else {
+                ans = abbreviation_recursion(memo, visited, s, t, i + 1, j);
+            }
+        }
+    }
+    (*memo)[i][j] = ans;
+    (*visited)[i][j] = true;
+    return (*memo)[i][j];
+}
+/**
+ *  (iterative dp function) https://www.hackerrank.com/challenges/abbr/problem:
+ *  Returns whether s can be converted to t with following rules:
+ *  a. Capitalize zero or more of a's lowercase letters from string s
+ *  b. remove all other lowercase letters from string s
```

Comment:
No need to add a link here. Consider adding a proper brief description of the function.
```suggestion
 * @brief Iterative DP function
 * @details
 * Returns whether s can be converted to t with the following rules:
 * a. Capitalize zero or more of `a`'s lowercase letters from string `s`
 * b. remove all other lowercase letters from string `s`
```